sbol2/document.py

```diff
@@ -447,39 +447,76 @@ def writeString(self):
         :param filename: The full name of the file you want to read
         (including file extension).
+        :param overwrite: Boolean indicating whether to overwrite existing objects
         :return: None
         """
-        self.logger.debug("Appending data from file: " + filename)
-        # Write our SBOL objects to an RDFlib graph
-        self.update_graph()
-        # Use rdflib to automatically merge the graphs together
-        self.graph.parse(filename, format="application/rdf+xml")
-        # Clear out the SBOL objects, but not the newly merged graph
-        self.clear(clear_graph=False)
-        # Base our internal representation on the new graph.
-        self.parse_all()
-
-    def appendString(self, sbol_str: str):
+        new_graph = rdflib.Graph()
+        new_graph.parse(filename, format='application/rdf+xml')
+        self._append_graph(new_graph, overwrite)
+
+        # self.logger.debug("Appending data from file: " + filename)
+        # # Write our SBOL objects to an RDFlib graph
+        # self.update_graph()
+        # # Use rdflib to automatically merge the graphs together
+        # self.graph.parse(filename, format="application/rdf+xml")
+        # # Clear out the SBOL objects, but not the newly merged graph
+        # self.clear(clear_graph=False)
+        # # Base our internal representation on the new graph.
+        # self.parse_all()
+
+    def appendString(self, sbol_str: str, overwrite=True):
         """
         Read an RDF/XML document from a string and attach the SBOL
         objects to this Document.
 
         New objects will be added to the existing contents of the Document.
         :param sbol_str: A string of RDF/XML
+        :param overwrite: Boolean indicating whether to overwrite existing objects
         :return: None
         """
-        # Save any changes we've made to the graph.
-        self.update_graph()
-        # Use rdflib to automatically merge the graphs together
-        self.graph.parse(data=sbol_str, format="application/rdf+xml")
-        # Base our internal representation on the new graph.
+        # ------------------------------------------------------------
+        # Load the new data into a graph
+        new_graph = rdflib.Graph()
+        new_graph.parse(data=sbol_str, format='application/rdf+xml')
+        self._append_graph(new_graph, overwrite)
+
+    def _append_graph(self, new_graph: rdflib.Graph, overwrite: bool = True):
+        # Gather all the objects that will be overwritten, stopping
+        # if the user says not to overwrite. If we clear as we go we lose
+        # the ability to find objects within objects. So gather the list
+        # here, and clear them as a second pass.
+        objects_to_clear = []
+        for s, _, _ in new_graph.triples((None, rdflib.RDF.type, None)):
+            existing_object = self.find(s)
+            if existing_object is not None:
+                if overwrite is False:
+                    return False
```

Comment:
This is probably fine for now, but at some point, we might want to add a more informative error message for the user, so they know which object is causing the conflict.  Also, you might want to perform this check after you've compiled _all_ the existing objects before failing, and then provide that to the user as an error.

Reply:
It sounds like you want an error raised if the user specifies `overwrite=False` and there is a conflict which prevents appending. I agree that communication from this method should be improved to tell the user whether the merge happened, and if not, why not. I'm not sure an exception is the right way, but an exception would force the user/programmer to deal with the situation, and ensure they are aware that the append did not happen.

Reply:
I think an exception is especially helpful when users are working interactively with the interpreter.  And it's not just about ensuring that they are aware that the append did not happen, but also about _why_ it did not happen, i.e. what is the object that is causing the collision.

Reply:
Proposal: if `overwrite=False` and there is a duplicate URI (thus requiring overwrite), then an `SBOLError` is raised with `error_code` `SBOLErrorCode.DUPLICATE_URI_ERROR` and `message` saying something like `"{URI} and N others would require overwriting". Sample message when merging "crispr_example.xml" into itself:

```
http://sbols.org/CRISPR_Example/CRPb_characterization_Circuit/CRISPR_Template/gRNA_b_map/1.0.0 and 106 others would require overwriting
```

Reply:
Cool.  Yes.  I like this!

Reply:
The code is in as of https://github.com/SynBioDex/pySBOL2/pull/373/commits/67f8762c5948f2d1dd45a4b66ba49c19892b900f to raise the exception as specified above.